### PR TITLE
Matcher's node implementation

### PR DESCRIPTION
**Pull request by rexagod**
@jywarren It's finally done! This PR has been worked on for about a week now and I'm just so glad to get this done. It supports user-specific image imports (the binding to "global" below) which is upped on a server then fetched by the html, which gets parsed for the results that are sent back to node.

test

I'm just bubbling with joy right now! Just let me know if everything looks good here so I can get to docs and the IS integration!

Thank you so much for supporting me throughout #9 (and generally too on a regular basis), much appreciated!

Also pinging @IshaGupta18 since she was interested in this PR's implementation.

**Edit**: For some reason the gif above had a delayed start, and is not infact the delay in the cli but the recorder itself.

### CODE REVIEW

**Comment by rexagod** [comment]
@jywarren This is good to go!!

**Comment by rexagod** [comment]
@sashadev-sky Sorry to have kept you waiting in LDI through the debugging process here. I'm 100% on it those LDI issues now!

**Comment by sashadev-sky** [comment]
Sorry for not helping with puppeteer!!! Completely meant to look at that. awesome job getting it done anyway 👍

**Comment by rexagod** [comment]
Haha no worries and thank you so much! Actually in the https://github.com/publiclab/matcher-core/pull/16/commits/c9defdf4121d15be1d5be5be344e2e51cfb25bf4 I had to redo almost everything from scratch that resulted in a large time reduction mostly owing to the shift from socket programming in earlier commits of this PR to EJS runtime generation in https://github.com/publiclab/matcher-core/pull/16/commits/c9defdf4121d15be1d5be5be344e2e51cfb25bf4 Coding was the easy part, however, formulating the workflow, without a doubt, wasn't.

https://github.com/publiclab/matcher-core/pull/16/commits/c9defdf4121d15be1d5be5be344e2e51cfb25bf4 took me two whole days to come up with this idea. EJS was my last option I could think of in hopes to get a 100% stable library, and it worked! Everything from here on out should not be time-consuming I suppose. 

Thank you all for your patience, really appreciate all this support that helped me get through this one! 🙌

cc @jywarren @sashadev-sky @sagarpreet-chadha @IgorWilbert @harshithpabbati

**Comment by rexagod** [comment]
PS. Just noting that this lib uses the heroku 16 stack i.e., `libxss1 libx11-xcb1 libxcomposite1 libxcursor1 libxdamage1 libxfixes3 libxi6 libxtst6 libnss3 libgconf-2-4 libxrandr2 libasound2 libatk1.0-0 libgtk2.0-0 libgtk-3-0 libxinerama1 libcairo-gobject2 ` libraries to work in a stable manner. Else for some reason it won't work in the first go. So I'll just drop this in the docs once this PR passes the code review.

**Comment by sashadev-sky** [comment]
@rexagod this is awesome! Should we be injecting our SVG in LDI this way too?

**Comment by rexagod** [comment]
Injecting SVGs into html templates? Can you elaborate on what ideas you have in mind about incorporating EJS in LDI? Maybe open a new issue over at LDI including the workflow? I'm excited to see how this works out! 🙌

**Comment by rexagod** [comment]
Also, @jywarren I'm thinking of releasing v1 after we get this one in. What do you think?

**Comment by sashadev-sky** [comment]
@rexagod see the milestones section next to PRs. @jywarren let’s all find a time to discuss it? We just have to be strict with breaking changes after version 1 so we need to finish code restructuring and some key APIs like allowing a mode selection. (Things like making a distortableCollection.edit and making sure actions are no longer in the global scope) Also finalizing how we will represent modes vs actions in a way that makes sense. And I guess all noticeable bugs should be fixed (keybindings, keymapper, EXIF, revert). 

I think we are really close though!! Just have to tie the loose ends and get our code to stability, I am finding myself making breaking changes more often than not still. 

I was also thinking converting to ES6 would be nice before release 1.

File: README.md [hunk 1]
```diff
@@ -57,6 +60,11 @@ This library takes a set of different options whose expanded map is provided bel
       }
     });
 ```
+- `browser`: NOT OPTIONAL. Set to `true` for regular in-browser usage. `false` is used for debug purposes (prints results in the console).
```
Comment (jywarren) [inline README.md:63]: Just checking - why can't this just default to true? No big deal, just curious!

File: README.md [hunk 2]
```diff
@@ -57,6 +60,11 @@ This library takes a set of different options whose expanded map is provided bel
       }
     });
 ```
+- `browser`: NOT OPTIONAL. Set to `true` for regular in-browser usage. `false` is used for debug purposes (prints results in the console).
+- `caching`: Enables cache mechanism for repetitive point detections. Defaults to `true`.
+- `leniency`: Minimum threshold value for which a point qualifies as a "match". Defaults to `30`%.
```
Comment (jywarren) [inline README.md:65]: Maybe clarify that it is a percentage, or offer a range of OK values?

File: index.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
+const chromeLauncher = require('chrome-launcher');
```
Comment (jywarren) [inline index.js:1]: Looks awesome! Will this be runnable in gh-pages?

**Review by jywarren — APPROVED** [review]

**Comment by jywarren** [comment]
Looks awesome, i'd love some short replies to a few questions, but after that we can merge this! Let's push it to gh-pages too please!

File: index.js (hunk 3 above)
Comment (rexagod) [reply index.js]: Done in #14 ! Please have a look.

File: README.md (hunk 1 above)
Comment (rexagod) [reply README.md:63]: Removed it altogether! No need for this now! Added a `query` option instead to satisfy issue #4

File: index.js (hunk 3 above)
Comment (rexagod) [reply index.js:1]: @jywarren Yes. The gh-pages is served from the browserified `orb.core.min.js` file generated from the src folder.

File: README.md (hunk 2 above)
Comment (rexagod) [reply README.md:65]: Done here https://github.com/publiclab/matcher-core/blob/add-readme/README.md#arguments

**Comment by rexagod** [comment]
@jywarren  Pushed @ https://publiclab.github.io/matcher-core/ 🎉

File: README.md (hunk 1 above)
Comment (rexagod) [reply README.md:63]: Latest readme can be seen in the add-readme branch

here -> https://github.com/publiclab/matcher-core/blob/add-readme/README.md#arguments

Additionally, I've clarified a bit further on the constructor signature here -> https://github.com/publiclab/matcher-core/blob/add-readme/README.md#browser

**Comment by rexagod** [comment]
Also note that the node logic was re-written after this PR so the constructor is even simpler now: `Promise.resolve(require('matcher-core')).then(fetchPoints);`. That's it! All this has been updated in the add-readme branch.

**Comment by jywarren** [comment]
That's awesome!!!

**Comment by rexagod** [comment]
@jywarren I've merged the add-readme branch that had all of the commits from the other PRs (this and #17 ) so closing this one!

**Pull request closed without merging**